blog/_posts/2018-04-18-scala-3.md

```diff
@@ -0,0 +1,104 @@
+---
+layout: blog-detail
+post-type: blog
+by: Martin Odersky
+title: "Towards Scala 3"
+---
+
+Now that Scala 2.13 is around the corner, it's time to consider the
+roadmap beyond it. It's been no secret that the work on Dotty over the
+last 5 years was intended to explore what a new Scala could look
+like. We are now at a stage where we can commit: Dotty will become
+Scala 3.0.
+
+Of course, this statement invites many follow-up questions. Here are
+some answers we can already give today. We expect there will be more
+questions and answers as things shape up.
+
+_When will it come out?_
+
+The intent is to publish the final Scala 3.0 soon after Scala 2.14. At the
+current release schedule (which might still change), that means late
+2019 or early 2020.
+
+_What is Scala 2.14 for?_
+
+Scala 2.14's main focus will be on smoothing the migration to Scala
+3. It will do this by defining migration tools, shim libraries, and
+targeted deprecations, among others.
+
+_What's new in Scala 3?_
+
+Scala has pioneered the fusion of object-oriented and functional
+programming in a typed setting. Scala 3 will be a big step towards
+realizing the full potential of these ideas. Its main objectives are
+to
+
+ - become more opinionated by promoting programming idioms we found
+to work well,
+ - simplify where possible,
+ - eliminate inconsistencies and surprising behavior,
+ - build on strong foundations to ensure the design hangs well together,
+ - remove some language constructs that proved to be not essential,
+ - include some new constructs that improve the language’s consistency, safety, ergonomics, and performance.
+
+The main language changes, either implemented or projected, are listed
+in the [Reference section on the Dotty website](http://dotty.epfl.ch/docs/reference/overview.html).
+Many of the new features will be submitted to the SIP process, subject to
+approval.
+
+_What about migration?_
+
+Scala 3 is neither source nor binary compatible with Scala 2. However:
+
+
+ - Scala 3 code can link against Scala 2 artifacts (Scala 2.12 and upwards).
+ - Scala 3 and Scala 2 share the same standard library.
+ - With some small tweaks it is possible to cross-build code for both Scala 2 and 3.
+   We will work on a guide defining the shared language subset that can be compiled under both versions.
+ - The Scala 3 compiler has a `-language:Scala2` option that lets it compile most Scala 2 code
+   and at the same time highlights necessary rewritings as migration warnings.
+ - The compiler can perform many of the rewritings automatically using a `-rewrite` option.
+ - Migration through automatic rewriting will also be offered through the `scalafix` tool,
+   which does not require Scala 3 to be installed.
```

Comment:
This seems peculiar and is perhaps an unnecessary level of detail. If I'm migrating to Scala 3 anyway, why do I care if this particular step doesn't require Scala 3 to be present?